### DISCUSSION TITLE
Trying to add support for (Tuya) MHCOZY TYZGTH16A DS18B20 temperature sensor with relay

### DISCUSSION THREAD

**Discussion by WayneManion** [Koenkk/zigbee2mqtt#23123] [Devices]
I bought a pretty unique Zigbee device I found on Amazon:  https://www.amazon.com/dp/B0D4DV81TH

It is essentially a Zigbee chip and a DS18B20 probe thermometer. The plan is to attach the probe to my dryer vent pipe to be able to determine when the dryer is running and when it is finished. I already have a YoLink sensor unit attached to the dryer vent and this method works very well. The YoLink setup is great, but it uses batteries and this Zigbee unit is mains-powered. I prefer mains-powered sensors when possible. This would also free up the YoLink sensor for some other job. 

I tried following the 'support a new Tuya device' instructions and was able to make some headway. I asked around on the Discord and was advised to try using the "modernExtends" format for device support. 

So far I have:
```const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    fingerprint: [
    {
      modelID: 'TS000F',
            manufacturerName: '_TZ3218_7fiyo3kv',
        },
    ],
    model: 'TYZGTH16A', 
    vendor: 'MHCOZY', 
    description: 'Temperature sensor controller with DS18B20 waterproof probe', 
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
     extend: [ tuya.modernExtend.tuyaMagicPacket(),
               tuya.modernExtend.tuyaOnOff(),
               tuya.modernExtend.dpTemperature({dp: 0x66, scale: 10}) ]
};

module.exports = definition;
```

I can see a reasonable-looking temperature value in the exposes tab of the Z2M web UI. I can also see a switch. When I press the button on the device the state of the switch in the web UI changes. 

When I try to change the state of the switch or change the value of the power-on behavior setting I get an error like this:
`Publish 'set' 'state' to 'Dryer vent' failed: 'Error: ZCL command 0xa4c1382c30531e81/0 genOnOff.off({}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":false,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Timeout - 14011 - 0 - 227 - 6 - 11 after 10000ms)'`

I don't really want to buy a Tuya Zigbee hub to snoop for the other parameters. I believe the device probably has some way to control the relay based on temperature probe input, but I don't see anything like that in the Z2M logs.

**Comment by WayneManion** [comment]
I bought a Tuya Zigbee hub and ran the steps in "Find the data points" instructions.

```
{
    "1": "Switch 1",
    "7": "Timer 1",
    "14": "Restart Status",
    "19": "Delay-off Schedule",
    "101": "Work Mode",
    "102": "Temp Current",
    "103": "Humidity Value",
    "104": "Maxhum Set",
    "105": "Minihum Set",
    "106": "Upper Temp",
    "107": "Lower Temp",
    "108": "Temp Correction",
    "109": "Humidity Calibration",
    "110": "Fault State",
    "111": "Child Lock",
    "112": "Hum Sensitivity",
    "113": "Temp Sensitivity",
    "114": "Alarm",
    "115": "sensor_dect",
    "116": "CF",
    "119": "switch_auto_1",
    "121": "温度告警",
    "122": "湿度告警",
    "123": "告警开关"
}
```
Google says:
121 tranlates to "temperature alarm"
122 translates to "humidity alarm"
123 translates to "alarm switch" #

**Comment by WayneManion** [comment]
From the Tuya debug logs:

Switch 1: Off and On
Work mode: auto and Manual
Temp current: values in degrees C
Timer 1: I couldn't get any values to appear here
Humidity value: no humidity sensor, no values
Maxhum set: no humidity sensor, no values
Minihum set: no humidity sensor, no values
Upper temp: I couldn't get any values to appear here, even when I set the device to auto
Lower temp: I couldn't get any values to appear here, even when I set the device to auto
Temp correction: values in degrees C
Humidity calibration: -1% (I set this value in the app. My device has no humidity sensing capabilities)
Restart status: Power-off and Restart Memory 
Fault state: I couldn't get any values to appear here 
Child lock: ON or OFF
Hum sensitivity: no humdiity sensor, but I could set values in the app and see them online
Temp sensitivity: values like 0.50* C as set in the app (the app offered options 0.1-1.0)
Delay-off schedule: AQA8 when I set the delay in the app to one minute, AQB4 when set to two minutes, AQB7 when set to 2:03, AQ27 when set to 58 minutes, AA2Y when disabled
sensor_dect: 'none' when I unplugged the sensor from the main box, tpm when I plugged it back in while the main box was plugged into mains
CF: I couldn't get any values to appear here, even when I changed the reporting units from C to F in the app
Alarm: I couldn't get any values to appear here

switch_auto_1: (control mode could only be set to temperature in the app) 
examples:
Temp >30C turn on, <25C turn off: 80002C0100000101FA00000000
Turn >31C turn on, <25C turn off: 8000360100000101FA00000000	
Turn >30C turn on, <24C turn off: 80002C0100000101F000000000

Temp >30C turn off, <25C turn on: 80002C0100000001FA00000001
Turn >31C turn off, <25C turn on: 8000360100000101FA00000001
Turn >30C turn off, <24C turn on: 80002C0100000101F000000001

**Comment by gianlucaf81** [comment]
I'm following carefully, my TYZGTH16A (Vine program) is also arriving, I'm really curious to understand how to add an unsupported device.

**Comment by gianlucaf81** [comment]
hi, trying to add custom .js from interface I receive this error: Request 'zigbee2mqtt/bridge/request/options' failed with error: 'external_converters must be array'

- solution -

The JS file must be inside the /opt/zigbee2mqtt/data/ folder (where the zigbee2mqtt configuration.yaml file is located). The following code should be added to the configuration.yaml file

```
external_converters:
  - TS000F.js
```

**Comment by gianlucaf81** [comment]
uh... here we go
![image](https://github.com/Koenkk/zigbee2mqtt/assets/39622729/42505741-9788-4f4a-b6f3-7beecd3485e2)

**Reply by emylive** [reply]
Hi, how did you add it?

**Reply by gianlucaf81** [reply]
> Hi, how did you add it?

i used this guide on yt watch?v=WV-S4jTBVMc but i use z2m in lxc version on proxmox so as i wrote above the folder where to put the js file is not the one in the video

**Reply by emylive** [reply]
Thank you very much! it's seems like you 
![Capture d’écran 2024-06-28 161432](https://github.com/Koenkk/zigbee2mqtt/assets/16337316/9f906dab-9fd3-4da2-90b6-b9ec4f1fa3ba)

**Reply by gianlucaf81** [reply]
Are you used WayneManion configuration at the beginning of the discussion?

**Reply by WayneManion** [reply]
> Are you used WayneManion configuration at the beginning of the discussion?

I can tell that emylive is not using my partially-working config because mine does not have battery or humidity. The module plugs in to mains power. The sensor with mine is a DS18B20 and measures temperature only.

**Reply by emylive** [reply]
Here is my device [https://fr.aliexpress.com/item/1005007040718331.html?spm=a2g0o.order_detail.order_detail_item.3.2e4f7d56vmMOTH&gatewayAdapt=glo2fra#nav-specification](url)
Now it works with your configuration thanks a lot!

![Capture d’écran 2024-06-29 152550](https://github.com/Koenkk/zigbee2mqtt/assets/16337316/4e4ecc3c-2cc9-403f-80fb-bad5edbefdc0)


```const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    fingerprint: [
    {
      modelID: 'TS000F',
            manufacturerName: '_TZ3218_ya5d6wth',
        },
    ],
    model: 'TS000F', 
    vendor: '_TZ3218_ya5d6wth', 
    description: 'Temperature sensor controller with DS18B20 waterproof probe', 
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
     extend: [ tuya.modernExtend.tuyaMagicPacket(),
               tuya.modernExtend.tuyaOnOff(),
               tuya.modernExtend.dpTemperature({dp: 0x66, scale: 10}) ]
};


module.exports = definition;

**Comment by brianacox** [comment]
I've just literally received one of these devices.... been trying to add it to ZHA and seen this post.... anyone succeeded ?   any advice (novice here) welcomed....

**Comment by Ricklws** [comment]
If anyone needs a little more help, after playing with the unit I was able to teak the code so that you can now see all of the relays.

![image](https://github.com/user-attachments/assets/38033bf8-8115-4699-96a3-bacab182911c)

**Reply by brianacox** [reply]
Fantastic..... I will try this....   I'm a bit of a newbie ... I've got a /homeassistant/custom_zha_quirks folder and I've created a file with the code you've provided and named it TS000F.py ..... and then deleted the device in ZHA and added it back in.... but to no avail.   I did change the vendor name to vendor: '_IZ3218_7fiyo3kv'   ... anything else that I need to do?

**Reply by brianacox** [reply]
Or is this code just for Z2M?  Not ZHA?

**Comment by jobabi** [comment]
Hi everybody,
I would also like to use such a sensor. could that device be integrated into the supported devices within the next updates ?
best wishes
Jochen

**Comment by WayneManion** [comment]
I added the code submitted by emylive. The temperature probe appears to work, but I have a lot of errors in my logs:

```
z2m: Failed to configure 'Dryer vent', attempt 3 (Error: ZCL command 0xa4c1382c30531e81/0 genBasic.read(["manufacturerName","zclVersion","appVersion","modelId","powerSource",65534], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"reservedBits":0,"writeUndiv":false}) failed (Timeout - 29286 - 0 - 6 - 0 - 1 after 10000ms) at Object.start (/app/node_modules/zigbee-herdsman/src/utils/waitress.ts:59:23) at ZStackAdapter.sendZclFrameToEndpointInternal (/app/node_modules/zigbee-herdsman/src/adapter/z-stack/adapter/zStackAdapter.ts:598:47) at processTicksAndRejections (node:internal/process/task_queues:95:5) at Queue.execute (/app/node_modules/zigbee-herdsman/src/utils/queue.ts:36:20) at Request.send (/app/node_modules/zigbee-herdsman/src/controller/helpers/request.ts:82:20) at Endpoint.zclCommand (/app/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:911:28) at Endpoint.read (/app/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:443:29) at configureMagicPacket (/app/node_modules/zigbee-herdsman-converters/src/lib/tuya.ts:418:9) at Object.configure (/app/node_modules/zigbee-herdsman-converters/src/index.ts:190:21) at Configure.configure (/app/lib/extension/configure.ts:125:13))
```

I'm not sure about the changes RIcklws added. As far as I can tell there is only one relay in the device.

**Reply by brianacox** [reply]
Hi Wayne.... will it now work for ZHA too?   automatically, or presumably I will need to update HA for 2004.9.0 ?

**Reply by WayneManion** [reply]
I would be shocked if this worked at all for ZHA. This repo is for Zigbee2MQTT. All Discussions/Issues relate to Zigbee2MQTT.

**Comment by jobabi** [comment]
After the update the device is working - thank you
I have also technical question here. Did anyone tried to connect a second temperature sensor ?

I used a tasmota device via WLAN and here it was possible to connect multiple sensors to the same GPIO Pin.
I ve got then kind of automatically Temp1, temp, temp3.....
